{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import scipy as sp\n",
    "import pandas as pd\n",
    "from scipy import stats\n",
    "import matplotlib.pyplot as plt\n",
    "import time\n",
    "import amt.method as md\n",
    "import amt.data_loader as dl\n",
    "import pickle\n",
    "\n",
    "%matplotlib inline\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load the data "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/martin/anaconda3/lib/python3.6/site-packages/IPython/core/interactiveshell.py:2698: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n"
     ]
    }
   ],
   "source": [
    "with open('../../parkinsons/parkinsons.pickle', 'rb') as f:\n",
    "    X = pickle.load(f)\n",
    "    y = pickle.load(f)\n",
    "    miss_prop = pickle.load(f)\n",
    "y = y-1\n",
    "file_map = '../../parkinsons/parkinsons.map'\n",
    "df_map = pd.read_csv(file_map, delimiter='\\t', \n",
    "                     names=['chromosome', 'snp', 'start', 'end'])\n",
    "n_sample, n_snp = X.shape\n",
    "ind_snp = np.array(miss_prop<0.05, dtype=bool)\n",
    "n_hypothesis = np.sum(ind_snp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute the expected observations\n",
    "Exp = np.zeros([8, n_snp], dtype=float)\n",
    "for iy in range(2):\n",
    "    for ix in range(4):\n",
    "        Exp[iy*4+ix,:] = np.mean(y==iy) * np.mean(X==ix,axis=0)\n",
    "Exp = Exp*n_sample\n",
    "r_Exp = 1/Exp.clip(min=1e-6)*(Exp>0)\n",
    "chi2_obs = md.compute_chi2(y, X, Exp, r_Exp)\n",
    "data_gwas = {'X':X, 'y':y, 'Exp':Exp, 'r_Exp':r_Exp, 'chi2_obs':chi2_obs}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Result analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "304\n"
     ]
    }
   ],
   "source": [
    "snp_list = ['rs10501570', 'rs281357', 'rs2242330', 'rs1480597', 'rs6826751', 'rs4888984',\n",
    "            'rs4862792', 'rs3775866', 'rs2235617', 'rs988421', 'rs7097094', 'rs999473',\n",
    "            'rs1912373', 'rs1887279', 'rs2986574', 'rs11090762', 'rs6125829', 'rs7796855',\n",
    "            'rs355477', 'rs3010040', 'rs2296713', 'rs355461', 'rs355506', 'rs355464',\n",
    "            'rs1497430', 'rs11946612']\n",
    "# result_file = '/home/martin/adapative_MC_test/results/GWAS/GWAS_amt_alpha_0.005_rep1.pickle'\n",
    "# result_file = '/home/martin/adapative_MC_test/results/GWAS/GWAS_amt_alpha_0.05_rep0.pickle'\n",
    "result_file = '../../results/GWAS/GWAS_amt_alpha_0.1_rep0.pickle'\n",
    "with open(result_file, 'rb')as f:\n",
    "    res_AMT = pickle.load(f)\n",
    "h_amt = res_AMT['h_amt']\n",
    "p_amt_ub = res_AMT['p_hat_ub']\n",
    "p_amt_lb = res_AMT['p_hat_lb']\n",
    "p_amt = res_AMT['p_hat']\n",
    "print(np.sum(h_amt))\n",
    "# print(np.arange(h_amt.shape[0])[h_amt])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "######################\n",
      "rs10501570\n",
      "       chromosome         snp  start       end\n",
      "262094         11  rs10501570      0  84095494\n",
      "decision [ True]\n",
      "miss_prop=0.0019, p_amc= [-9.59697468e-06] [6.93645165e-05] [3.8033274e-06]\n",
      "\n",
      "######################\n",
      "rs281357\n",
      "       chromosome       snp  start       end\n",
      "343235         17  rs281357      0  19683106\n",
      "decision [ True]\n",
      "miss_prop=0.0000, p_amc= [-6.5849564e-06] [7.24366922e-05] [9.21916018e-06]\n",
      "\n",
      "######################\n",
      "rs2242330\n",
      "       chromosome        snp  start       end\n",
      "100847          4  rs2242330      0  68276015\n",
      "decision [ True]\n",
      "miss_prop=0.0000, p_amc= [-6.5849564e-06] [7.24366922e-05] [9.21916018e-06]\n",
      "\n",
      "######################\n",
      "rs1480597\n",
      "       chromosome        snp  start       end\n",
      "237247         10  rs1480597      0  44481115\n",
      "decision [ True]\n",
      "miss_prop=0.0223, p_amc= [-2.11240548e-06] [7.17599404e-05] [1.52353239e-05]\n",
      "\n",
      "######################\n",
      "rs6826751\n",
      "       chromosome        snp  start       end\n",
      "100837          4  rs6826751      0  68262621\n",
      "decision [ True]\n",
      "miss_prop=0.0019, p_amc= [-4.22742011e-06] [7.24666375e-05] [1.2570302e-05]\n",
      "\n",
      "######################\n",
      "rs4888984\n",
      "       chromosome        snp  start       end\n",
      "337424         16  rs4888984      0  78066835\n",
      "decision [ True]\n",
      "miss_prop=0.0000, p_amc= [-6.11804747e-06] [7.42744641e-05] [1.03715552e-05]\n",
      "\n",
      "######################\n",
      "rs4862792\n",
      "       chromosome        snp  start        end\n",
      "115749          4  rs4862792      0  188576499\n",
      "decision [ True]\n",
      "miss_prop=0.0483, p_amc= [-1.1884073e-05] [6.9268264e-05] [0.]\n",
      "\n",
      "######################\n",
      "rs3775866\n",
      "       chromosome        snp  start       end\n",
      "100842          4  rs3775866      0  68272946\n",
      "decision [ True]\n",
      "miss_prop=0.0000, p_amc= [6.16466027e-06] [6.99994703e-05] [2.47068821e-05]\n",
      "\n",
      "######################\n",
      "rs2235617\n",
      "       chromosome        snp  start       end\n",
      "380211         20  rs2235617      0  47988384\n",
      "decision [ True]\n",
      "miss_prop=0.0093, p_amc= [1.57909457e-05] [7.26270915e-05] [3.50755211e-05]\n",
      "\n",
      "######################\n",
      "rs988421\n",
      "      chromosome       snp  start       end\n",
      "10385          1  rs988421      0  72261857\n",
      "decision [ True]\n",
      "miss_prop=0.0019, p_amc= [1.66980003e-05] [7.43839519e-05] [3.6407503e-05]\n",
      "\n",
      "######################\n",
      "rs7097094\n",
      "       chromosome        snp  start       end\n",
      "237252         10  rs7097094      0  44530696\n",
      "decision [ True]\n",
      "miss_prop=0.0000, p_amc= [3.91164096e-05] [7.36579147e-05] [5.37423788e-05]\n",
      "\n",
      "######################\n",
      "rs999473\n",
      "       chromosome       snp  start       end\n",
      "237250         10  rs999473      0  44502322\n",
      "decision [ True]\n",
      "miss_prop=0.0000, p_amc= [3.77838853e-05] [7.38236479e-05] [5.28944658e-05]\n",
      "\n",
      "######################\n",
      "rs1912373\n",
      "       chromosome        snp  start       end\n",
      "258213         11  rs1912373      0  56240441\n",
      "decision [ True]\n",
      "miss_prop=0.0000, p_amc= [3.36586783e-05] [7.46120179e-05] [5.02628824e-05]\n",
      "\n",
      "######################\n",
      "rs1887279\n",
      "      chromosome        snp  start        end\n",
      "22113          1  rs1887279      0  180641817\n",
      "decision [ True]\n",
      "miss_prop=0.0000, p_amc= [2.4722479e-05] [7.44711264e-05] [4.33594147e-05]\n",
      "\n",
      "######################\n",
      "rs2986574\n",
      "      chromosome        snp  start        end\n",
      "22108          1  rs2986574      0  180638271\n",
      "decision [ True]\n",
      "miss_prop=0.0019, p_amc= [3.39640985e-05] [7.50595938e-05] [5.06393834e-05]\n",
      "\n",
      "######################\n",
      "rs11090762\n",
      "       chromosome         snp  start       end\n",
      "407979         22  rs11090762      0  46075844\n",
      "decision [False]\n",
      "miss_prop=0.0056, p_amc= [0.05928677] [0.55813087] [0.23]\n",
      "\n",
      "######################\n",
      "rs7796855\n",
      "       chromosome        snp  start       end\n",
      "175197          7  rs7796855      0  49434707\n",
      "decision [ True]\n",
      "miss_prop=0.0000, p_amc= [7.11592306e-05] [7.11592306e-05] [7.113449e-05]\n",
      "\n",
      "######################\n",
      "rs355477\n",
      "       chromosome       snp  start       end\n",
      "100830          4  rs355477      0  68225291\n",
      "decision [False]\n",
      "miss_prop=0.0074, p_amc= [9.09531752e-05] [9.09531752e-05] [9.0928435e-05]\n",
      "\n",
      "######################\n",
      "rs3010040\n",
      "      chromosome        snp  start        end\n",
      "22111          1  rs3010040      0  180639879\n",
      "decision [ True]\n",
      "miss_prop=0.0000, p_amc= [5.33640831e-05] [7.51970624e-05] [6.33536807e-05]\n",
      "\n",
      "######################\n",
      "rs2296713\n",
      "      chromosome        snp  start        end\n",
      "22112          1  rs2296713      0  180641374\n",
      "decision [ True]\n",
      "miss_prop=0.0000, p_amc= [5.33640831e-05] [7.51970624e-05] [6.33536807e-05]\n",
      "\n",
      "######################\n",
      "rs355461\n",
      "       chromosome       snp  start       end\n",
      "100828          4  rs355461      0  68209490\n",
      "decision [ True]\n",
      "miss_prop=0.0000, p_amc= [7.49943074e-05] [7.49943074e-05] [7.49695668e-05]\n",
      "\n",
      "######################\n",
      "rs355506\n",
      "       chromosome       snp  start       end\n",
      "100829          4  rs355506      0  68214848\n",
      "decision [ True]\n",
      "miss_prop=0.0000, p_amc= [7.49943074e-05] [7.49943074e-05] [7.49695668e-05]\n",
      "\n",
      "######################\n",
      "rs355464\n",
      "       chromosome       snp  start       end\n",
      "100825          4  rs355464      0  68207890\n",
      "decision [False]\n",
      "miss_prop=0.0112, p_amc= [0.00010715] [0.00018058] [0.00013918]\n",
      "\n",
      "######################\n",
      "rs1497430\n",
      "       chromosome        snp  start       end\n",
      "100821          4  rs1497430      0  68186580\n",
      "decision [ True]\n",
      "miss_prop=0.0037, p_amc= [3.79200941e-05] [7.20256989e-05] [5.23281057e-05]\n",
      "\n",
      "######################\n",
      "rs11946612\n",
      "       chromosome         snp  start       end\n",
      "100819          4  rs11946612      0  68164737\n",
      "decision [ True]\n",
      "miss_prop=0.0037, p_amc= [3.74282315e-05] [7.33307323e-05] [5.24701786e-05]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# df_map_c4 = df_map.loc[df_map['chromosome']==4]\n",
    "n_overlap = 0\n",
    "temp_df_map = df_map.loc[ind_snp]\n",
    "for snp in snp_list:\n",
    "    temp = np.array(temp_df_map['snp']==snp, dtype=bool)\n",
    "    if np.sum(temp)>0:\n",
    "        print('######################')\n",
    "        print(snp)\n",
    "        print(temp_df_map.loc[temp])\n",
    "        print('decision', h_amt[temp])\n",
    "        print('miss_prop=%0.4f, p_amc='%(miss_prop[ind_snp][temp]),\n",
    "                                         p_amt_lb[temp], p_amt_ub[temp], p_amt[temp])\n",
    "        if h_amt[temp][0]:\n",
    "            n_overlap+=1\n",
    "        print('')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
